### DISCUSSION TITLE
Redaxo Composer Setup - Beispiel?

### DISCUSSION THREAD

**Discussion by Sysix** [FriendsOfREDAXO/.github#4] [Q&A] [answered]
Hallo Jungs, Mädels und Co.,

ich hab die Frage schon mehrmals gestellt und nie eine brauchbare Antwort bekommen. 
Diesmal Versuch Nummer 3 :)

## Anforderungen

Ich möchte gerne Redaxo Projekte über Composer verwalten, hierzu soll folgendes Möglich sein:

- Redaxo Version in Composer auswählen
- Addon und dessen Versionen  in Composer auswählen
- Redaxo Core und Addon Code soll nicht in Git auftauchen

## Testprojekt

Am Anfang bin ich schon gescheitert an das normale Setup.  
Irgendwann kam mir `$REX['PATH_PROVIDER']` und `$REX['URL_PROVIDER']` in den Augen.

Daraus ist dieses kleine Testprojekt enstanden:
https://github.com/Sysix/redaxo-composer

## Probleme

- (Fast) Alle AddOns haben keine composer.json und publizieren diese auch in der packagist.org
- Updatefähigkeit ist noch nicht direkt möglich. Mann muss das Plugin und dessen alte Version kennen um ein Update auszuführen
- Teile der Sfotware ist nicht mit unterschiedlichen `$REX['PATH_PROVIDER']` und `$REX['URL_PROVIDER']` getestet worden

-----

## Frage

- Wie habt ihr euer Composer Setup? 
- Gibt es vielleicht ein Repo als Beispiel?
- Sind meine Anforderungen überhaupt aktuell möglich?
- Wird eine Composer Kompatibilität angestrebt?

**Comment by eaCe** [comment]
Ich kann dir die Frage nicht beantworten, aber vielleicht einen kleinen Teil beitragen.
Angefangen hatte ich REDAXO via Composer zu nutzen um Unit-Tests für Addons zu schreiben. Dazu habe ich nur den REDAXO-Core benötigt. An diesen kommt man problemlos über Composer.
Für das Setup habe ich mir die Dateien aus dem Core geklaut:
https://github.com/redaxo/redaxo/blob/main/.tools/bin/setup
https://github.com/redaxo/redaxo/blob/main/.tools/bootstrap.php

Um Addons zu installieren nutze ich dann die Konsole:
z.B. `vendor/redaxo/source/redaxo/bin/console package:install 'phpmailer'`
https://redaxo.org/doku/main/console

**Comment by gharlan** [comment, marked as answer]
Ich kann dir hier auch keine zufriedenstellende Antwort geben. Ich kann dir nur sagen, dass es auch unser großer Wunsch ist, dass sich Core und Addons normal über Composer laden lassen. Und es ist auch ein Thema, worüber ich schon lange nachdenke, wie ich das gescheit umsetze. Denke auch, dass das kommen wird, aber ich kann keinen Zeithorizont nennen.

> Teile der Sfotware ist nicht mit unterschiedlichen $REX['PATH_PROVIDER'] und $REX['URL_PROVIDER'] getestet worden

Dazu kurz als Hinweis, dass wir bei Yakamara in unseren Projekten immer einen angepassten Path-Provider haben und das zumindest bei uns zuverlässig funktioniert. Url-Provider (also die öffentlichen URLs) belassen wir aber bei den Defaults.
Unsere Ordnerstruktur kann man [hier](https://github.com/yakamara/yak#konsole) sehen. Uns ging es da vor allem auch um den public-Ordner. In `public/redaxo` liegt dabei dann nur noch die index.php, damit das Backend weiter über `/redaxo` läuft. Aber die src-Dateien liegen im Root-src-Ordner.

**Reply by Sysix** [reply]
> Url-Provider (also die öffentlichen URLs) belassen wir aber bei den Defaults

Genau da hab ich die ersten Probleme erkannt. Es wird z.B. im Backend nach Zugriffsrechten von Ordner überprüft (via Ajax).

> Unsere Ordnerstruktur kann man [hier](https://github.com/yakamara/yak#konsole) sehen. 

Danke für den netten Einblick. Leider entspricht das nicht meinen Vorstellungen. (vorallem bezüglich Composer).

---

Ich glaube als Erstes sollte man die Update Funktion von Composer ermöglichen.  
Mein erster Plan wäre, die AddOns und Core Versionen in einer Datenbank zu speichern und ab nach einen Login vom Benutzer, der via Nutzerrecht hat, die DB Update durchzuführen. 

Gerne darf auch auch Konsole Befehl erstellt werden, der im CI/CD Prozess ausgeführt werden kann.

Danach sollten die AddOns eine composer.json im Git besitzen und Redaxo am Besten ein eigenes Composer-Repo via [satis](https://github.com/composer/satis), wo die AddOns hochgeladen werden :) 

https://wpackagist.org/ zieht sich z.B. von der offiziellen Seite die Versionen. Vielleicht wollt ihr sowas ähnliches bereitstellen ;)

PS: Ich schließe das Thema, da es wahrscheinlich ein langer Weg bis dahin ist. Danke für eine ehrliche Antwort
